crates/devolutions-pedm/src/db/libsql.rs

```diff
@@ -178,24 +203,400 @@ impl Database for LibsqlConn {
+                                return Err(DbError::InvalidEnum(InvalidEnumError {
+                                    value: n,
+                                    enum_name: "ElevationKind",
+                                }))
+                            }
+                        },
+                        target_must_be_signed: row.get::<bool>(5)?,
+                    },
+                    users: Vec::new(),
+                });
+            }
+
+            if let Some(a) = current_assignment.as_mut() {
+                let user_id: Option<i64> = row.get(6)?;
+
+                if let Some(_) = user_id {
+                    a.users.push(User {
+                        account_name: row.get(7)?,
+                        domain_name: row.get(8)?,
+                        account_sid: row.get(9)?,
+                        domain_sid: row.get(10)?,
+                    });
+                }
+            }
+        }
+
+        if let Some(a) = current_assignment {
+            assignments.push(a);
+        }
+
+        Ok(assignments)
+    }
+
+    async fn get_assignment(&self, profile: &Profile) -> Result<Assignment, DbError> {
+        let mut users = Vec::new();
+
+        let mut rows = self
+            .query(
+                "
+                SELECT
+                    u.account_name,
+                    u.domain_name,
+                    u.account_sid,
+                    u.domain_sid
+                FROM policy pol
+                JOIN user u ON u.id = pol.user_id
+                WHERE pol.profile_id = ?1
+                ",
+                params![profile.id],
+            )
+            .await?;
+
+        while let Some(row) = rows.next().await? {
+            users.push(User {
+                account_name: row.get(0)?,
+                domain_name: row.get(1)?,
+                account_sid: row.get(2)?,
+                domain_sid: row.get(3)?,
+            });
+        }
+
+        Ok(Assignment {
+            profile: profile.clone(),
+            users,
+        })
+    }
+
+    async fn set_assignments(&self, profile_id: i64, users: Vec<User>) -> Result<(), DbError> {
+        let mut tx: Transaction = self.transaction().await?;
+
+        tx.execute("DELETE FROM policy WHERE profile_id = ?1", params![profile_id])
+            .await?;
+
+        for user in users {
+            let user_id = self.get_or_insert_user(&mut tx, &user).await?;
+
+            tx.execute(
+                "INSERT INTO policy (profile_id, user_id) VALUES (?1, ?2)",
+                params![profile_id, user_id],
+            )
+            .await?;
+        }
 
         tx.commit().await?;
         Ok(())
     }
 
+    async fn set_user_profile(&self, user: &User, profile_id: Option<i64>) -> Result<(), DbError> {
+        // `unwrap` should be sound, the user should not be able to select a profile unless they already exist in the database
+        let user_id = self.get_user_id(user).await?.unwrap();
```

Comment:
issue: This is true, but I also don’t really agree that it’s fine for the same reasons we can’t trust client-side forms to never send malformed data to the server in web pages. Granted here everything is local and the worst that would happen is the agent process crashing. If you did that to avoid dealing with the error handling, I think it’s fine for the V1, but add a `FIXME` comment so I can look into that later :slightly_smiling_face:

Reply:
You're absolutely right and this was pure laziness on my part for not wanting to add a new error type. I've done that now.